### ISSUE TITLE
LightGBM Booster MIP produces wrong predictions on sparse features due to threshold/FeasibilityTol mismatch

### ISSUE THREAD

**Issue by harshaamzn** [Gurobi/gurobi-machinelearning#496]
Summary
When LightGBM trains on a feature that has many exact-zero values (common in sparse fraud features, NLP, categorical embeddings, etc.), it may emit split thresholds of approximately ±1e-35. These represent the data distribution's natural "zero vs non-zero" boundary.

gurobi_ml.lightgbm.add_lgbm_booster_constr encodes these as indicator constraints like leaf_active = 1 -> x <= 1e-35. Gurobi's default FeasibilityTol is 1e-6. So for any value of x in roughly [0, 1e-6], the constraint x - 1e-35 <= 0 is treated as satisfied within tolerance — activating the wrong leaf.

LightGBM at inference uses exact comparison: for x = 1e-10, 1e-10 > 1e-35 routes right, but the MIP's indicator constraint for x <= 1e-35 treats x = 1e-10 as satisfying it. The MIP returns a score that disagrees with the actual LightGBM model.

This cannot be worked around by lowering FeasibilityTol — Gurobi's minimum is 1e-9, which is still much larger than the 1e-35 threshold used by LightGBM.

Environment
lightgbm==4.6.0
gurobi-machinelearning==1.5.6
gurobipy==13.0.2
Python 3.13.12, Linux x86_64
Minimal reproducer

import numpy as np
import lightgbm as lgb
import gurobipy as gp
from gurobipy import GRB
from gurobi_ml.lightgbm import add_lgbm_booster_constr

# Training data: feature 0 has many EXACT zeros + some non-zero active values
np.random.seed(42)
N = 5000
X = np.random.randn(N, 5).astype(np.float32)
X[:, 0] = 0.0
active = np.random.rand(N) < 0.05
X[active, 0] = np.random.choice([-5, 5], size=active.sum()).astype(np.float32)
y = (X[:, 0] != 0).astype(np.int32)

model = lgb.train(
    {'objective': 'binary', 'num_leaves': 4, 'max_depth': 2,
     'learning_rate': 0.3, 'verbose': -1, 'seed': 42,
     'deterministic': True, 'min_data_in_leaf': 10},
    lgb.Dataset(X, y), num_boost_round=1,
)

# Build MIP
env = gp.Env()
m = gp.Model(env=env); m.Params.OutputFlag = 0
x_vars = m.addMVar(shape=5, lb=-10.0, ub=10.0, name='x')
y_var = m.addVar(lb=-GRB.INFINITY, ub=GRB.INFINITY, name='score')
add_lgbm_booster_constr(m, model, x_vars, y_var)
m.update()

# Compare MIP vs actual model predictions
for x0 in [0.0, 1e-35, 1e-10, 1e-6, 1e-3, 1.0, 5.0]:
    x_test = np.array([x0, 0.0, 0.0, 0.0, 0.0], dtype=np.float32)
    lgbm_score = float(model.predict(x_test.reshape(1, -1), raw_score=True)[0])
    for j in range(5):
        x_vars[j].LB = float(x_test[j])
        x_vars[j].UB = float(x_test[j])
    m.setObjective(y_var, GRB.MINIMIZE)
    m.optimize()
    mip_score = float(y_var.X)
    print(f"x[0]={x0:+.0e}: lgbm={lgbm_score:+.4f} mip={mip_score:+.4f} diff={lgbm_score-mip_score:+.2e}")
Observed output

x[0]=+0e+00: lgbm=-3.2727 mip=-3.2727 diff=+9.72e-08   # OK
x[0]=+1e-35: lgbm=-3.2727 mip=-3.2727 diff=+9.72e-08   # OK
x[0]=+1e-10: lgbm=+3.1157 mip=-3.2727 diff=+6.39e+00   # MISMATCH
x[0]=+1e-06: lgbm=+3.1157 mip=-3.2727 diff=+6.39e+00   # MISMATCH
x[0]=+1e-03: lgbm=+3.1157 mip=+3.1157 diff=-3.99e-08   # OK
x[0]=+1e+00: lgbm=+3.1157 mip=+3.1157 diff=-3.99e-08   # OK
x[0]=+5e+00: lgbm=+3.1157 mip=+3.1157 diff=-3.99e-08   # OK
For x[0] in roughly (1e-35, 1e-6) (the range between LightGBM's actual threshold and Gurobi's FeasibilityTol), the MIP returns the wrong leaf. The discrepancy is 6.39 in log-odds — enough to flip a fraud detection decision.

Impact
This matters for counterfactual use cases (documented as a primary use case of gurobi-machinelearning). When searching for the minimum perturbation that flips a decision, the MIP's "solution" may not actually flip the real model, because the MIP silently routed small feature values to the wrong leaf.

In our production fraud model (LightGBM 500 trees, ~200 features with many sparse feature columns), ~20% of test inputs show MIP-vs-actual score discrepancies of 0.005–0.05 in log-odds, and counterfactual searches routinely return X values where the real LightGBM model has NOT flipped.

Expected behavior
The MIP's y_var should equal model.predict(x_test, raw_score=True) at any fixed x_test, to machine precision.

Suggested fix
When add_lgbm_booster_constr reads a LightGBM tree and encounters a split threshold smaller than the solver's feasibility tolerance, it should either:

Normalize the threshold to a value that separates the "zero zone" from active values within the solver's tolerance (e.g., use epsilon parameter to explicitly shift the threshold by some safe margin, then compensate on the other side).
Warn the user that the threshold is below tolerance and the MIP may disagree with the actual model in a specific region.
Document that this can happen when LightGBM trains on sparse features, and recommend preprocessing (replace 0 values with something outside the threshold range) before building the MIP.
Option 1 is the most robust but requires careful handling to avoid breaking tight-threshold models.

Workaround we're using
We've implemented a custom MIP formulation based on Mišić (2020, "Optimization of Tree Ensembles") that uses split-indicator binaries per unique threshold rather than per-leaf indicator constraints. This formulation uses only linear constraints (no indicator constraints), so Gurobi's FeasibilityTol does not interfere with the threshold comparisons. MIP score matches LightGBM score to machine precision in our tests.

We would prefer to use the library add_lgbm_booster_constr directly for our production system if this issue is resolved.

**Comment by pobonomo** [comment]
Hi,

Thanks for the report.

On the basic issue you are right and it is a fundamental issue. For training and evaluating LightGBM would only need to do comparisons using floating point algorithms while Gurobi builds the solution through an involved numerical procedure that has looser tolerances.

Note that there is an epsilon parameter to enforce a tighter separation. There is a discussion there https://gurobi-machinelearning.readthedocs.io/en/stable/user/mip-models.html#gradient-boosting-regression in our documentation on how to set it.

By default, we chose to set it to 0 and this induces what you see. Our thinking was that if there is an epsilon perturbation that changes the outcome anyway the classification is not very robust. The drawback of increasing epsilon is that you create some holes in your domains and it may even become infeasible.

For your use case, I suppose you want to set this epsilon. If, I would want to be as safe as possible to avoid mismatches in prediction, I would:

- set Gurobi tolerances to a tight value (maybe not 1e-9 but 1e-8). Also tighten the Integer [Infeasibility tolerance](https://docs.gurobi.com/projects/optimizer/en/current/reference/parameters.html#intfeastol)
- Set the epsilon in add_lgbm_booster_constr to something larger (at least twice as large)
- Also set the parameter [PreSOS1BigM](https://docs.gurobi.com/projects/optimizer/en/current/reference/parameters.html#parameterintfeastol) to 0. This one will cost probably in terms of solver performance. You could also try it independently of the two above.

All this should reduce the number of mismatch but of course you may miss some counterfactual examples.

About your first suggestion, I don't think that we can detect this. As far as I understand, your argument assumes that the values are non-negative but we don't necessarily know it at build time.

What I see we could clearly do is show a warning in every cases.

I remember reading that paper you mention a few years ago. I am not sure that I see right now why it fixes the issue. I think we could think about switching to that formulation.

If you can try my suggestions above, I'd be curious to know if they improve the situation.

**Comment by harshaamzn** [comment]
Thanks Pierre — appreciate the quick and thorough response.

We tried the epsilon + tightened tolerances approach first, but ran into the domain-hole problem you mentioned: with epsilon > 0, some orders that are perfectly valid in the original LightGBM model become infeasible in the MIP (because the epsilon gap excludes their actual feature values from any leaf). For counterfactual search this is particularly problematic — it can make the MIP report "no flip exists" when one clearly does.

Our workaround that avoids domain holes:

```python
# In the constraint-building loop, clamp tiny thresholds:
SAFETY_FLOOR = 1e-5  # well above FeasibilityTol but below any "real" threshold
if abs(threshold) < SAFETY_FLOOR:
    threshold = SAFETY_FLOOR  # preserves the "zero vs non-zero" semantics
```

Rationale: LightGBM emits thresholds like 2.3e-35 to represent "is this feature (approximately) zero?" The exact value 2.3e-35 vs 1e-5 doesn't matter for the split semantics — both separate zero from non-zero. But 1e-5 is safely above Gurobi's FeasibilityTol, so the indicator constraint is numerically meaningful.

This only applies when `abs(threshold) < SAFETY_FLOOR`. For thresholds at normal scale (e.g., 0.5, 1.2, 100.0), nothing changes.

**On the Mišić formulation:** the reason it helps with this specific issue is that Mišić uses *split-indicator* variables shared across trees (one binary per unique threshold per feature), rather than per-leaf flow variables. When a threshold is clamped from 2.3e-35 to 1e-5, the split-indicator formulation naturally merges all near-zero thresholds on the same feature into a single binary — so the clamping has zero impact on the solution space. In the leaf formulation, clamping can subtly change which leaves are reachable because the flow constraints reference the threshold directly.

Happy to submit a PR with our threshold-clamping fix if that's useful. We've validated it on production-scale LightGBM models (3,400 trees, 1.7M splits) and confirmed that all counterfactuals match `model.predict()` after the fix.

Thanks again for maintaining this library — it's been essential for our work.

**Comment by pobonomo** [comment]
Now I think that I understand better your suggestion. This would amount to basically changing the thresholds in the tree is that correct?

I think that we could indeed introduce a minimum_threshold parameter. But I would probably keep it to 0 by default with a warning as you suggested.

If you have code ready, I am more than happy to look at a PR.

It's even nicer if you can include some tests with your use case and a bit of documentation.

Thanks!

**Comment by pobonomo** [comment]
Hi,
I don't know if you had a chance to make a PR. I thought that maybe I would be able to do something simple. I tried it in https://github.com/Gurobi/gurobi-machinelearning/pull/506

If you can try it, that would be great.

**Comment by harshaamzn** [comment]
Thanks Pierre. I will get back in a few days

**Comment by harshaamzn** [comment]
Thanks @pobonomo — PR #506 matches how we'd resolve it, and the design is good.

The root cause is purely the threshold-vs-tolerance mismatch: LightGBM emits ±1e-35 split thresholds on sparse (mostly-zero) features, and any threshold below Gurobi's FeasibilityTol (default 1e-6, floor 1e-9) makes the split constraint vacuous, so the MIP activates the wrong leaf while LightGBM uses exact comparison. We caught it not by inspecting gurobi_ml but by replaying the MIP's solution back through the actual LightGBM booster and finding the scores disagreed — then traced it to the tiny thresholds in the model dump vs. the solver tolerance.

By that criterion the safety_floor clamp is correct whenever safety_floor ≥ FeasibilityTol: your suggested safety_floor=1e-5 sits safely above both the default 1e-6 and the 1e-9 floor, so clamped splits become numerically enforceable and the MIP routes consistently with LightGBM. The UserWarning when abs(threshold) < FeasibilityTol is the most valuable part for us — it turns a silent wrong answer into a visible, actionable one.

Two small suggestions:

Document that safety_floor should be set ≥ FeasibilityTol to actually fix the problem — a value below the tolerance won't help.
Note that a fixed floor can distort models whose legitimate thresholds are genuinely below it (we only use it safely because our features are unit-normalized), which is a good reason to keep it opt-in as you've done.
With safety_floor=1e-5 this resolves the issue for our production LightGBM models. 👍

**Comment by pobonomo** [comment]
Thanks for the comments @harshaamzn 

I will soon include it in a new release. For now it can be tested in the alpha release:

```
pip install -i https://test.pypi.org/simple/ gurobi-machinelearning==1.6.0a1
```
